onewire.py

nklever opened the pull request:
Based on file components/sensors/onewire.py I wrote this file components/switch/onewire.py for all onewire switch sensors listed on [owfs family site](http://owfs.sourceforge.net/family.html). It's only tested with family 3A (DS2413) switch, but according to the [owfs DS24xx man pages](http://owfs.org/index.php?page=man-pages-ds24xx) all of the DS24xx switches are similar. This file is working only on an installed [OWFS](http://owfs.org/) filesystem. 

A onewire switch must be configured by **platform: onewire** and **mount_dir: /path/to/owfs-path** in a switch part of the configuration.

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

nklever committed:
onewire.py

Based on file components/sensors/onewire.py I wrote this file components/switch/onewire.py for all onewire switch sensors listed on [owfs family site](http://owfs.sourceforge.net/family.html). It's only tested with family 3A (DS2413) switch, but according to the [owfs DS24xx man pages](http://owfs.org/index.php?page=man-pages-ds24xx) all of the DS24xx switches are similar. This file is working only on an installed [OWFS](http://owfs.org/) filesystem. 

A onewire switch must be configured by **platform: onewire** and **mount_dir: /path/to/owfs-path** in a switch part of the configuration.

homeassistant commented:
Hi @nklever, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#14616).

Once you do that we will be able to review and accept this pull request.

Thanks!

nklever committed:
update onewire.py 

Update onewire.py according to hints of houndcl-bot

nklever committed:
Update onewire.py

Update onewire.py according to hints of houndcl-bot

syssi commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,135 @@
+        """Turn the switch on."""
+        with open(self._device_file, 'w') as ds_device_file:
+            ds_device_file.write("1")
+        self._state = True
```
Please introduce an update method which reads and updates the state. Don't set the state here.

syssi commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,135 @@
+    def is_on(self):
+        """Return true if switch is on."""
+        if self._state is None:
+            self._state = self._readState()
```
Avoid I/O operations here. Just return the state. Update the state at the update method.

syssi commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,135 @@
+        state = None
+        try:
+            state = self._read_value_raw() == DEVICE_SWITCH_ON
+        except ValueError:
```
When is this exception raised?

syssi commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,135 @@
+        """Initialize the OneWire switch."""
+        self._name = name or DEVICE_DEFAULT_NAME
+        self._device_file = device_file
+        self._state = self._readState()
```
I/O isn't allowed here. Please set the state to `None`.

syssi commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,135 @@
+    @property
+    def should_poll(self):
+        """tbd, if control of OneWire Switch is not only done by hass"""
+        return False
```
True?

syssi commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,135 @@
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        with open(self._device_file, 'w') as ds_device_file:
+            ds_device_file.write("1")
```
You could use your const.

nklever committed:
Updated onewire.py

Updated according to Sebastian's hints

nklever committed:
Update onewire.py

Update onewire.py according to houndcl-bot

nklever committed:
Update onewire.py

Updated because of a foolish copy and paste error

nklever committed:
Update onewire.py

Updated according to a missing doc string

syssi commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,143 @@
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        if self._write_value_raw(DEVICE_SWITCH_ON):
+            self._state = True
```
HA calls the update method right after turn_{on,off}. Is this temporary fake state  really needed? I assume it can be removed.

nklever committed:
Update onewire.py

Update according to Sebastian's hint.
Thanks again for your hints !

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,141 @@
+    def update(self):
+        """Get the latest data from the switch and update the state."""
+        self._state = self._read_value_raw() == DEVICE_SWITCH_ON
+        return self._state
```
Nothing is checking this return value.

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,141 @@
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        if self._write_value_raw(DEVICE_SWITCH_ON):
+            self.schedule_update_ha_state()
```
Remove this too. It will be called automatically after the service call, since this is a polling entity.

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,141 @@
+    def turn_off(self, **kwargs):
+        """Turn the device off."""
+        if self._write_value_raw(DEVICE_SWITCH_OFF):
+            self.schedule_update_ha_state()
```
Same as above.

OttoWinter commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,141 @@
+
+    def __init__(self, name, device_file):
+        """Initialize the OneWire switch."""
+        self._name = name or DEVICE_DEFAULT_NAME
```
A default name is already defined by `switch_id_complete` - so `... or DEVICE_DEFAULT_NAME` doesn't do anything here. Please remove it.

OttoWinter commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,141 @@
+            msg = "1Wire switch file not found: %s", self._device_file
+            _LOGGER.warning(msg)
+        except OSError:
+            _LOGGER.warning("Error reading switch file %s", self._device_file)
```
You should probably `return None` in these except blocks. Otherwise, `content` will be undefined and below line will trigger another exception -> not good

OttoWinter commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,141 @@
+                content = ds_device_file.read()
+        except FileNotFoundError:
+            msg = "1Wire switch file not found: %s", self._device_file
+            _LOGGER.warning(msg)
```
Please just write

```python
_LOGGER.warning("1Wire switch file not found: %s", 
                self._device_file)
```

It's cleaner.

smurfix replied:
Also, the above doesn't work, you 'd need `warning(*msg)`.

OttoWinter commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,141 @@
+        """Write the value to the switch."""
+        if value not in [DEVICE_SWITCH_ON, DEVICE_SWITCH_OFF]:
+            _LOGGER.error("Error in setting wrong value %s", value)
+        else:
```
You could save a level of indentation by returning False here instead of doing the ugly `else:`. It would also help with the weird `_LOGGER.warning(msg)`

```python
if value not in [DEVICE_SWITCH_ON, DEVICE_SWITCH_OFF]:
    _LOGGER.error("Error in setting wrong value %s", value)
    return False
# ...
```

nklever committed:
Update onewire.py

Updating onewire.py according to the hints of Martin and Otto.
Thanks for this hints !

nklever committed:
Update onewire.py

... according to houndci-bot

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -125,17 +121,12 @@ def is_on(self):
         """Return true if switch is on."""
         return self._state
 
-    def update(self):
```
Don't remove `update`. This is needed to read the updated state. Just don't return anything from `update`.

nklever replied:
Hi Martin,
I've accidentally removed update ! 
Thanks for your request - I've commited it again with the previous update function.
Sincerely,
Nik

MartinHjelmare reviewed (changes requested):
Don't remove `update`. This is needed to read the updated state. Just don't return anything from `update`.

nklever committed:
Update onewire.py

Updating because of accidentally removing of update.
Thanks to Martin !

MartinHjelmare commented:
There's now a python wrapper for onewire OWFS C-API:
https://github.com/kipe/python-onewire

This is also published on pypi.

Maybe this should be extended to support switches and then we can use that library here?

nklever committed:
Update onewire.py

Thanks, Martin for your additional hints !

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,137 @@
+        """Write the value to the switch."""
+        if value not in [DEVICE_SWITCH_ON, DEVICE_SWITCH_OFF]:
+            _LOGGER.error("Error in setting wrong value %s", value)
+            return None
```
Just `return`.

MartinHjelmare replied:
This still needs addressing.

nklever commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,140 @@
+        """Write the value to the switch."""
+        if value not in [DEVICE_SWITCH_ON, DEVICE_SWITCH_OFF]:
+            _LOGGER.error("Error in setting wrong value %s", value)
+            return False
```
Yes, you are right - but then I should add an else-Block here, which was in contrast to Otto Winters [comment](https://github.com/home-assistant/home-assistant/pull/14616#discussion_r190989821)

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,137 @@
+    device_names = config.get(CONF_NAMES, {})
+
+    for family_file_path in glob(os.path.join(base_dir, '*', 'family')):
+        with open(family_file_path, "r") as family_file:
```
Here are more double quotes unexpectedly.

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,137 @@
+            switch_id = os.path.split(
+                os.path.split(family_file_path)[0])[1]
+            for switch_key, switch_value in DEVICE_SWITCHES[family].items():
+                switch_id_complete = switch_id + "_" + switch_key
```
Here as well.

nklever commented:
Yes, seems interesting - I will check it. My proposal here is based on the current components.sensor.onewire.py file - which should then also be changed.

nklever committed:
Update onewire.py

again thanks to Martin's hints !

MartinHjelmare commented on homeassistant/components/switch/onewire.py:
```diff
@@ -0,0 +1,137 @@
+                            self._device_file)
+        except OSError:
+            _LOGGER.warning("Error reading switch file %s", self._device_file)
+        return
```
This should still return `None`. I think you removed  `None` at the wrong line.

nklever replied:
Yes, you are right - I took the wrong line !

nklever committed:
Update onewire.py

... taken the wrong line !
Thanks, Martin !

MartinHjelmare commented:
Codewise this now looks good to me. But I think we need a decision on if we should continue to add onewire platforms without a standalone interface library. I'd like at least a couple of member's to say yea or nay to the question: Should we merge now, without standalone library?

nklever commented:
I think your rule to add devices only with standalone libraries seems reasonable - I will try to extend the python-onewire library, test it and change both homeassistant onewire files to be based on this library. But it may take a time !

fabaff commented:
This PR seems to gone stale. Please re-open it when you want to proceed.
